### PR TITLE
Adding FQDN filtering into Juniper SRX generator.

### PR DESCRIPTION
**Pull request by ankenyr**
This completes the work to resolve https://github.com/aerleon/aerleon/issues/253
SRX will now be able to filter FQDN's in their terms.

FQDN's can be used in a term by using either `destination-fqdn` or `source-fqdn` the below is an example

```
  - name:  allow_dns
    destination-fqdn: CLOUDFLARE_DNS
    action: accept
```

See https://github.com/aerleon/aerleon/pull/291 for how to add FQDN's to the addressbook.

### CODE REVIEW

File: aerleon/lib/addressbook.py [hunk 1]
```diff
@@ -4,12 +4,55 @@
...
+                yield zone, group, ips, fqdns
+
+    def GetFQDN(self, zone: str, name: str):
+        return self._fqdn_addressbook[zone][name]
+
+    def GetAddressTokensInZone(self, zone: str):
+        return [i for i in self._ip_addressbook['zone']]
+
+    def GetFQDNTokensInZone(self, zone: str):
+        return [i for i in self._fqdn_addressbook['zone']]
```
Comment (jtwb) [inline aerleon/lib/addressbook.py:45]: Looks like a typo. There should probably be a quick test that covers these functions.

```suggestion
    def GetAddressTokensInZone(self, zone: str):
        return [i for i in self._ip_addressbook[zone]]

    def GetFQDNTokensInZone(self, zone: str):
        return [i for i in self._fqdn_addressbook[zone]]
```

Comment (jtwb) [reply aerleon/lib/addressbook.py:45]: Also, could this just be `return copy(self._fqdn_addressbook[zone])` ?

File: aerleon/lib/junipersrx.py [hunk 2]
```diff
@@ -529,8 +537,8 @@ def _TranslatePolicy(self, pol: policy.Policy, exp_info: int):
                 # SRX policies are controlled by addresses that are used within, so
                 # policy can be at the same time inet and inet6.
                 if self._GLOBAL_ADDR_BOOK in self.addr_book_type:
-                    for zone in self.addressbook.addressbook:
-                        for _, ips in sorted(self.addressbook.addressbook[zone].items()):
+                    for zone in self.addressbook._ip_addressbook:
+                        for _, ips in sorted(self.addressbook._ip_addressbook[zone].items()):
```
Comment (jtwb) [inline aerleon/lib/junipersrx.py:541]: Should this code use one of the new addressbook methods?

File: aerleon/lib/cisco.py [hunk 3]
```diff
@@ -208,9 +208,10 @@ def __str__(self) -> str:
         # for using cisco, which has decided to implement its own meta language.
 
         # Create network object-groups
-        for name, ips in self.addressbook.addressbook[''].items():
+        for zone, name, ips, _ in self.addressbook.WalkAddressBook():
             for version in (4, 6):
                 vips = [i for i in ips if i.version == version]
+                vips = nacaddr.CollapseAddrList(vips)
```
Comment (jtwb) [inline aerleon/lib/cisco.py:214]: Is the addition of CollapseAddrList here related to the FQDN / AddressBook change?

File: aerleon/lib/addressbook.py [hunk 4]
```diff
@@ -4,12 +4,55 @@
...
 class Addressbook:
     def __init__(self) -> None:
-        self.addressbook = collections.OrderedDict()
+        self._fqdn_addressbook = collections.OrderedDict()
+        self._ip_addressbook = collections.OrderedDict()
+
+    def GetZoneNames(self) -> List:
+        return list(self._ip_addressbook.keys())
+
+    def WalkAddressBook(self, zone_filter=None):
```
[the remark this answers is not in the document]
Comment (ankenyr) [reply aerleon/lib/addressbook.py:19]: Done

File: aerleon/lib/addressbook.py (hunk 1 above)
Comment (ankenyr) [reply aerleon/lib/addressbook.py:45]: Done

File: aerleon/lib/junipersrx.py (hunk 2 above)
Comment (ankenyr) [reply aerleon/lib/junipersrx.py:541]: naaaa

**Review by jtwb — APPROVED** [review]

**Pull request merged**